chainerx_cc/chainerx/routines/linalg.cc

```diff
@@ -98,4 +98,83 @@ Array Dot(const Array& a, const Array& b, nonstd::optional<Dtype> out_dtype) {
     return out_matrix.Reshape(out_shape);
 }
 
+Array Solve(const Array& a, const Array& b) {
+    if (a.ndim() != 2) {
+        throw DimensionError{"ChainerX solve supports only 2-dimensional arrays."};
+    }
+    if (a.shape()[0] != a.shape()[1]) {
+        throw DimensionError{"Matrix is not square."};
+    }
+    CheckEqual(a.device(), b.device());
+    CheckEqual(a.dtype(), b.dtype());
+    Dtype dtype = internal::GetMathResultDtype(b.dtype());
+    Array out = Empty(b.shape(), dtype, b.device());
+
+    {
+        NoBackpropModeScope scope{};
+        a.device().backend().CallKernel<SolveKernel>(a, b, out);
+    }
+
+    // Reference:
+    // https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
+    // Sec. 2.3.1 Matrix inverse product
+    {
+        BackwardBuilder bb{"solve", {a, b}, out};
+        if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+            bt.Define([out_tok = bb.RetainOutput(0), a_tok = bb.RetainInput(0), a_dtype = a.dtype()](BackwardContext& bctx) {
+                const Array& a = bctx.GetRetainedInput(a_tok);
+                const Array& out = bctx.GetRetainedOutput(out_tok);
+                const Array& gout = *bctx.output_grad();
+                bctx.input_grad() =
+                        -Dot(Solve(a.Transpose(), gout).At(std::vector<ArrayIndex>{Slice{}, NewAxis{}}),
```

Comment:
Can you use `chainerx::ExpandDims` instead of `chainerx::At` with `chainerx:NewAxis`?

Reply:
Sure, I didn't know about `chainerx::ExpandDims`.